File: day02/src/main.rs
use std::fs;

fn main() {
    let content = fs::read_to_string("./day02/inp");
    match content {
        Ok(content) => {
            let reports = into_matrix(&content);
            println!("{:?}", task1(reports.clone()));
            println!("{:?}", task2(reports.clone()));
        }
        Err(e) => {
            println!("{:?}", e);
        }
    }
}

fn task1(reports: Vec<Vec<i64>>) -> i64 {
    let mut safe: i64 = 0;

    for report in reports {
        if is_safe(report){
            safe += 1;
        }
    }

    safe
}

fn task2(reports: Vec<Vec<i64>>) -> i64 {
    let mut safe: i64 = 0;

    for report in reports {
        let mut was_safe: bool = false;

        for mutation in get_mutations(report) {
            if is_safe(mutation){
                was_safe = true;
                break;
            }
        }

        if was_safe{
            safe += 1;
        }
    }

    safe
}

fn into_matrix(content: &str) -> Vec<Vec<i64>> {
    content.lines()
        .map(|line| {
            line.split_whitespace()
                .filter_map(|word| word.parse::<i64>().ok())
                .collect::<Vec<i64>>()
        })
        .collect()
}

fn is_safe(report: Vec<i64>) -> bool {
    let old_dir = to_dir(report[0], report[1]);

    if old_dir == ERR_UNSAFE {
        return false;
    }

    for i in 0..report.len() - 1 {
        if check(report[i], report[i + 1], old_dir) == ERR_UNSAFE {
            return false;
        }
    }

    true
}

fn check(a: i64, b: i64, old_dir: &str) -> &str {
    let dir = to_dir(a, b);

    if !(1..=3).contains(&(a-b).abs()) || dir == ERR_UNSAFE || old_dir != dir{
        return ERR_UNSAFE
    }

    return ""
}

const ERR_UNSAFE: &str = "ERR_UNSAFE";

fn to_dir(a: i64, b: i64) -> &'static str {
   match a.cmp(&b) {
        std::cmp::Ordering::Equal => ERR_UNSAFE,
        std::cmp::Ordering::Greater => "decr",
        std::cmp::Ordering::Less => "incr",
    }
}

fn get_mutations(report: Vec<i64>) -> Vec<Vec<i64>> {
    let mut out = Vec::new();

    out.push(report.clone());

    for i in 0..report.len() {
        let mut mutation = report.clone();
        mutation.remove(i);
        out.push(mutation);
    }

    out
}

File: day02/main.go
package main

import (
	_ "embed"
	"errors"
	"log/slog"
	"strconv"
	"strings"

	"github.com/solokirrik/aoc2024/utils"
)

//go:embed inp
var inp string

var errUnsafe = errors.New("line is unfase")

const (
	incr = "incr"
	decr = "decr"
)

func main() {
	inplines := strings.Split(inp, "\n")

	slog.Info("Part 1:", "safe:", part1(inplines))
	slog.Info("Part 2:", "safe:", part2(inplines))
}

func part1(inpLines []string) int {
	safe := 0

	for _, line := range inpLines {
		report := getReport(line)

		if err := isSafe(report); err != nil {
			continue
		}

		safe++
	}

	return safe
}

func part2(inpLines []string) int {
	safe := 0

	for _, line := range inpLines {
		wasSafe := false
		report := getReport(line)

		for _, mut := range mutations(report) {
			if err := isSafe(mut); err == nil {
				wasSafe = true
				break
			}
		}

		if wasSafe {
			safe++
		}
	}

	return safe
}

func getReport(inp string) []int {
	chars := strings.Split(inp, " ")
	report := make([]int, 0, len(chars))

	for i := range chars {
		c1, _ := strconv.Atoi(chars[i])
		report = append(report, c1)
	}

	return report
}

func mutations(report []int) [][]int {
	mutations := [][]int{report}

	for i := range len(report) {
		option := make([]int, 0, len(report)-1)

		for j := range len(report) {
			if i != j {
				option = append(option, report[j])
			}
		}

		mutations = append(mutations, option)
	}

	return mutations
}

func isSafe(report []int) error {
	oldDir, err := toDir(report[0], report[1])
	if err != nil {
		return err
	}

	for i := 0; i < len(report)-1; i++ {
		if err := check(report[i], report[i+1], oldDir); err != nil {
			return err
		}
	}

	return nil
}

func toDir(a, b int) (string, error) {
	switch {
	case a == b:
		return "", errUnsafe
	case a > b:
		return decr, nil
	case a < b:
		return incr, nil
	default:
		return "", nil
	}
}

func check(a, b int, oldDir string) error {
	if utils.Abs(a-b) < 1 || utils.Abs(a-b) > 3 {
		return errUnsafe
	}

	dir, err := toDir(a, b)
	if err != nil {
		return errUnsafe
	}

	if oldDir != dir {
		return errUnsafe
	}

	return nil
}
